File: ios/customer_io/Sources/customer_io/LiveActivities/CustomerIOLiveActivities.swift
import CioDataPipelines
import CioInternalCommon
import Flutter
import Foundation
#if canImport(CioLiveActivities)
import ActivityKit
import CioLiveActivities
import CioLiveActivities_Attributes
import CioLiveActivities_Templates
#endif

/// Flutter plugin sub-handler for Live Activities.
///
/// Gated on `canImport(CioLiveActivities)` and iOS 16.2+. Custom activity types work from Dart
/// through the SDK-owned `CIOCustomAttributes`: ActivityKit needs a concrete Swift type to register
/// an activity and observe its push-to-start token, and a metatype can't cross a method channel, so
/// the SDK owns the type and Dart supplies its values.
public class CustomerIOLiveActivities: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?

    #if canImport(CioLiveActivities)
    /// Reverse-DNS activity type identifiers for the SDK's built-in templates. These are the same
    /// strings the backend sends as `notificationType` and that Android's `LiveNotificationType`
    /// exposes, so Dart, both native SDKs, and the wire format share one vocabulary.
    enum TypeIdentifier {
        static let segments = "io.customer.livenotifications.segments"
        static let countdownTimer = "io.customer.livenotifications.countdowntimer"
        /// Discriminator Dart sends for the custom template. Not a wire identifier — the activity is
        /// reported under the app's own `liveNotifications.customType`.
        static let custom = "custom"
    }

    /// Type-erased handles keyed by activity id. The native `start` returns a generic
    /// `CIOLiveActivity<Attributes>` that can't be stored in a homogeneous map, so we keep closures
    /// that capture the concrete handle and rebuild its content-state from a Dart map on update.
    private struct ActivityBox {
        let update: ([String: Any]) async throws -> Void
        let end: ([String: Any]?) async throws -> Void
    }

    private var activities: [String: ActivityBox] = [:]
    private let lock = NSLock()

    /// Records an `update`/`end` aimed at an activity this process never started. Not surfaced to
    /// Dart — see the call sites — but worth a log, because the same message covers both a genuine
    /// caller mistake and the expected post-restart / push-started cases.
    private static func logUnknownActivity(_ activityId: String, method: String) {
        DIGraphShared.shared.logger.info(
            "Live Activities: \(method) ignored — no activity with id \(activityId) was started by this app session. " +
                "Activities started before an app restart, or started by a push, are not tracked in-process."
        )
    }
    #endif

    public static func register(with _: FlutterPluginRegistrar) {}

    /// Reports an `opened` metric for a tapped Live Activity and returns the deep link to route to.
    /// Call this from the app's `openURL` entry point so a Live Activity tap is attributed.
    ///
    /// - Returns: the customer's redirect URL for a Customer.io widget URL (`nil` when it carries
    ///   none), or `url` unchanged when it isn't a Customer.io URL — so existing routing still
    ///   handles non-CIO links.
    #if canImport(CioLiveActivities)
    @discardableResult
    public static func handleWidgetUrl(_ url: URL) -> URL? {
        CustomerIO.liveActivities.handleWidgetUrl(url)
    }
    #else
    public static func handleWidgetUrl(_ url: URL) -> URL? { url }
    #endif

    init(with registrar: FlutterPluginRegistrar) {
        super.init()

        let channel = FlutterMethodChannel(name: "customer_io_live_activities", binaryMessenger: registrar.messenger())
        methodChannel = channel
        registrar.addMethodCallDelegate(self, channel: channel)
    }

    deinit {
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
    }

    #if canImport(CioLiveActivities)
    /// Build the Live Activities module from the SDK config's `liveNotifications` key, for
    /// registration on the SDK config builder at init. Registers the enabled built-in template
    /// attribute types so `CustomerIO.liveActivities.start` can request them, and so push-to-start
    /// tokens register for them at SDK init.
    ///
    /// Unrecognized type identifiers are ignored: a newer native SDK may ship templates this
    /// wrapper build doesn't know, and that must never break registration of the ones it does.
    static func module(from params: [String: AnyHashable]) -> LiveActivitiesModule? {
        guard let liveConfig = params["liveNotifications"] as? [String: AnyHashable] else { return nil }
        guard #available(iOS 16.2, *) else { return nil }
        let types = (liveConfig["types"] as? [String]) ?? []
        var builder = LiveActivityConfigBuilder()
        for type in types {
            switch type {
            case TypeIdentifier.segments:
                builder = builder.register(CIOSegmentsAttributes.self)
            case TypeIdentifier.countdownTimer:
                builder = builder.register(CIOCountdownTimerAttributes.self)
            default:
                continue
            }
        }
        // The custom template registers one SDK-owned Swift type under the app's own identifier.
        // That indirection is what lets a Dart app have a custom activity at all: the SDK needs a
        // metatype to register and to observe push-to-start for, and a metatype can't cross a
        // method channel.
        //
        // Nothing is cached from this: `start` learns that a custom activity isn't registered from
        // the nil handle the SDK returns, which stays correct if the type was registered elsewhere.
        if let customType = (liveConfig["customType"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
            !customType.isEmpty {
            builder = builder.register(CIOCustomAttributes.self, identifier: customType)
        }
        return LiveActivitiesModule(config: builder.build())
    }
    #endif

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "start":
            start(args, result: result)

        case "update":
            update(args, result: result)

        case "end":
            end(args, result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Bridge methods

    private func start(_ args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(CioLiveActivities)
        guard #available(iOS 16.2, *) else {
            return result(Self.unavailableError())
        }
        guard let map = args["payload"] as? [String: Any], let type = map["type"] as? String else {
            return result(FlutterError(code: "live_activity_start_failed", message: "payload.type is required", details: nil))
        }
        do {
            // A nil handle means the module isn't registered or this type wasn't enabled. The
            // native SDK logs and returns nil rather than throwing, so surface it as a
            // FlutterError — never a crash.
            let id: String
            switch type {
            case TypeIdentifier.segments:
                guard let handle = try CustomerIO.liveActivities.start(
                    CIOSegmentsAttributes(header: try Self.requireString(map, "header")),
                    contentState: try Self.segmentsState(from: map)
                ) else {
                    return result(Self.notRegisteredError(type))
                }
                store(handle: handle, contentBuilder: Self.segmentsState)
                id = handle.id
            case TypeIdentifier.countdownTimer:
                guard let handle = try CustomerIO.liveActivities.start(
                    CIOCountdownTimerAttributes(header: try Self.requireString(map, "header")),
                    contentState: try Self.countdownState(from: map)
                ) else {
                    return result(Self.notRegisteredError(type))
                }
                store(handle: handle, contentBuilder: Self.countdownState)
                id = handle.id
            case TypeIdentifier.custom:
                // No pre-check on the configured identifier: an unregistered type already comes back
                // as a nil handle, and that path is correct however the SDK was initialized — a
                // check against wrapper-captured config would refuse a type registered elsewhere.
                guard let handle = try CustomerIO.liveActivities.start(
                    CIOCustomAttributes(),
                    contentState: try Self.customState(from: map)
                ) else {
                    return result(Self.customNotRegisteredError())
                }
                store(handle: handle, contentBuilder: Self.customState)
                id = handle.id
            default:
                // A newer native SDK may know this type even though this wrapper build doesn't.
                // Fail softly so an unrecognized template can never crash the host app.
                return result(FlutterError(
                    code: "live_activity_type_unsupported",
                    message: "Unsupported Live Activity template: \(type)",
                    details: nil
                ))
            }
            result(id)
        } catch {
            result(FlutterError(code: "live_activity_start_failed", message: error.localizedDescription, details: nil))
        }
        #else
        result(Self.unavailableError())
        #endif
    }

    private func update(_ args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(CioLiveActivities)
        guard let activityId = args["activityId"] as? String else {
            return result(FlutterError(code: "live_activity_update_failed", message: "activityId is required", details: nil))
        }
        lock.lock()
        let box = activities[activityId]
        lock.unlock()
        // An unknown id means the update did not happen, so it must not report success. Unlike `end`
        // — where re-ending an already-ended activity is a legitimate no-op — there is nothing
        // idempotent about an update that never applied. Android *does* perform it (it routes the id
        // straight to the native SDK), so resolving here would make the same call report success on
        // both platforms while only one of them did anything.
        guard let box = box else {
            Self.logUnknownActivity(activityId, method: "update")
            return result(FlutterError(
                code: "live_activity_update_failed",
                message: "No live activity found for id \(activityId). On iOS only activities started " +
                    "in this app session can be updated.",
                details: nil
            ))
        }
        guard let map = args["payload"] as? [String: Any] else {
            return result(FlutterError(code: "live_activity_update_failed", message: "payload is required", details: nil))
        }
        Task {
            do {
                try await box.update(map)
                // FlutterResult must be invoked on the platform (main) thread; the Task
                // resumes off-main after the await, so hop back before replying.
                await MainActor.run { result(true) }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "live_activity_update_failed", message: error.localizedDescription, details: nil))
                }
            }
        }
        #else
        result(Self.unavailableError())
        #endif
    }

    private func end(_ args: [String: Any], result: @escaping FlutterResult) {
        #if canImport(CioLiveActivities)
        guard let activityId = args["activityId"] as? String else {
            return result(FlutterError(code: "live_activity_end_failed", message: "activityId is required", details: nil))
        }
        lock.lock()
        let box = activities[activityId]
        lock.unlock()
        // Unknown/already-ended id is treated as success (idempotent end). See `update` for why an
        // unknown id is not an error.
        guard let box = box else {
            Self.logUnknownActivity(activityId, method: "end")
            return result(true)
        }
        let map = args["payload"] as? [String: Any]
        Task {
            // FlutterResult must be invoked on the platform (main) thread; the Task
            // resumes off-main after the await, so hop back before replying.
            do {
                try await box.end(map)
                // Dropped only once the end succeeded. Removing up front would lose the handle on a
                // throw, and the retry would then take the unknown-id path above and report success
                // for an activity still on screen.
                self.forget(activityId)
                await MainActor.run { result(true) }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "live_activity_end_failed", message: error.localizedDescription, details: nil))
                }
            }
        }
        #else
        result(Self.unavailableError())
        #endif
    }

    // MARK: - Helpers

    #if canImport(CioLiveActivities)
    @available(iOS 16.2, *)
    private func store<A: ActivityAttributes>(
        handle: CIOLiveActivity<A>,
        contentBuilder: @escaping ([String: Any]) throws -> A.ContentState
    ) {
        let box = ActivityBox(
            // ActivityKit keeps the last content-state on screen when `end` is given none, so a
            // final payload is what lets the activity show a terminal state rather than freezing
            // mid-progress.
            update: { map in await handle.update(try contentBuilder(map)) },
            end: { map in await handle.end(try map.map(contentBuilder)) }
        )
        lock.lock()
        activities[handle.id] = box
        lock.unlock()
    }

    private func forget(_ activityId: String) {
        lock.lock()
        activities.removeValue(forKey: activityId)
        lock.unlock()
    }

    /// Thrown for a missing required field so the caller sees an error naming it, rather than an
    /// activity rendering with a blank line. Matches Android, whose `requireString` / `requireInt`
    /// already throw — the same call must not produce a silent half-rendered card on one platform
    /// and an error on the other.
    private struct MissingFieldError: LocalizedError {
        let field: String
        var errorDescription: String? { "\(field) is required" }
    }

    private static func requireString(_ map: [String: Any], _ key: String) throws -> String {
        guard let value = map[key] as? String else { throw MissingFieldError(field: key) }
        return value
    }

    private static func requireInt(_ map: [String: Any], _ key: String) throws -> Int {
        guard let value = map[key] as? NSNumber else { throw MissingFieldError(field: key) }
        return value.intValue
    }

    @available(iOS 16.2, *)
    private static func segmentsState(from map: [String: Any]) throws -> CIOSegmentsAttributes.ContentState {
        CIOSegmentsAttributes.ContentState(
            status: try requireString(map, "status"),
            substatus: map["substatus"] as? String,
            segmentsTotal: try requireInt(map, "segmentsTotal"),
            segmentsComplete: try requireInt(map, "segmentsComplete"),
            trailingText: map["trailingText"] as? String
        )
    }

    @available(iOS 16.2, *)
    private static func countdownState(from map: [String: Any]) throws -> CIOCountdownTimerAttributes.ContentState {
        var endTime: EpochSecondsDate?
        if let seconds = map["endTime"] as? NSNumber {
            endTime = EpochSecondsDate(Date(timeIntervalSince1970: seconds.doubleValue))
        }
        return CIOCountdownTimerAttributes.ContentState(
            title: try requireString(map, "title"),
            statusMessage: map["statusMessage"] as? String,
            endTime: endTime
        )
    }

    /// Builds the custom template's content-state from the payload's `data` map.
    ///
    /// Values are coerced to strings rather than rejected: a Dart `int`/`double`/`bool` arrives as
    /// `NSNumber`, and refusing them would make `{'eta': 5}` fail for no reason a caller can see.
    /// Anything without a sensible text form is dropped instead of stringifying as gibberish.
    @available(iOS 16.2, *)
    private static func customState(from map: [String: Any]) throws -> CIOCustomAttributes.ContentState {
        let raw = map["data"] as? [String: Any] ?? [:]
        var data: [String: String] = [:]
        for (key, value) in raw {
            switch value {
            case let string as String: data[key] = string
            case let number as NSNumber: data[key] = number.stringValue
            default: continue
            }
        }
        return CIOCustomAttributes.ContentState(data: data)
    }
    #endif

    private static func unavailableError() -> FlutterError {
        FlutterError(
            code: "live_activity_module_unavailable",
            message: "Live Activities require iOS 16.2 or later.",
            details: nil
        )
    }

    private static func notRegisteredError(_ type: String) -> FlutterError {
        FlutterError(
            code: "live_activity_type_not_registered",
            message: "Live Activity type '\(type)' is not registered. Add it to `liveNotifications.types` " +
                "in your Customer.io SDK config, and make sure your widget extension renders it.",
            details: nil
        )
    }

    /// The custom template's version of [notRegisteredError]: `type` is only the discriminator, so
    /// naming it would tell the caller nothing about what to fix.
    private static func customNotRegisteredError() -> FlutterError {
        FlutterError(
            code: "live_activity_type_not_registered",
            message: "No custom Live Activity type is registered. Set `liveNotifications.customType` " +
                "in your Customer.io SDK config to your own reverse-DNS identifier, and render " +
                "CIOCustomAttributes in your Widget Extension.",
            details: nil
        )
    }
}

File: android/src/main/kotlin/io/customer/customer_io/liveactivities/CustomerIOLiveActivities.kt
package io.customer.customer_io.liveactivities

import android.graphics.Color
import io.customer.customer_io.bridge.NativeModuleBridge
import io.customer.customer_io.bridge.nativeMapArgs
import io.customer.customer_io.utils.getAs
import io.customer.messagingpush.MessagingPushModuleConfig
import io.customer.messagingpush.ModuleMessagingPushFCM
import io.customer.messagingpush.di.pushModuleConfig
import io.customer.messagingpush.livenotification.LiveNotificationAsset
import io.customer.messagingpush.livenotification.LiveNotificationBranding
import io.customer.messagingpush.livenotification.LiveNotificationData
import io.customer.messagingpush.livenotification.LiveNotificationType
import io.customer.sdk.CustomerIOBuilder
import io.customer.sdk.core.di.SDKComponent
import io.customer.sdk.core.util.Logger
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter module implementation for Live Activities / Live Notifications.
 *
 * Live Notifications are hosted by the FCM push module ([ModuleMessagingPushFCM]); this bridge maps
 * the wrapper's template payloads to [LiveNotificationData] and forwards them at runtime. The module
 * *config* (enabled templates, custom types, branding) is applied onto the same
 * [MessagingPushModuleConfig] the push module builds (see [applyLiveActivitiesConfig]); this handler
 * therefore does not add a module of its own.
 */
// Public so host apps can register a custom Live Notification render callback via
// [setLiveNotificationCallback]; the constructor stays internal (the plugin owns instantiation).
class CustomerIOLiveActivities internal constructor(
    pluginBinding: FlutterPlugin.FlutterPluginBinding,
) : NativeModuleBridge, MethodChannel.MethodCallHandler {
    override val moduleName: String = "LiveActivities"
    override val flutterCommunicationChannel: MethodChannel =
        MethodChannel(pluginBinding.binaryMessenger, "customer_io_live_activities")
    private val logger: Logger = SDKComponent.logger

    // Live Notifications are hosted by the FCM push module. Reach it via the module registry
    // (the SDK-internal MODULE_NAME constant is not accessible, so use the literal value).
    //
    // Deliberately not runCatching: `as?` yields null instead of throwing, so a failure branch keyed
    // on a thrown error would never run and the message below would never be logged.
    private fun getPushModule(): ModuleMessagingPushFCM? {
        val module = SDKComponent.modules[PUSH_FCM_MODULE_NAME] as? ModuleMessagingPushFCM
        if (module == null) {
            logger.error("Live Notifications: push module is not initialized. Ensure the SDK is initialized with live activity templates enabled.")
        }
        return module
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // `start` is handled with the raw result rather than [nativeMapArgs] so it can report a
            // type that isn't enabled under the same error code iOS uses; see [start].
            "start" -> start(call, result)
            "update" -> call.nativeMapArgs(result, ::update)
            "end" -> call.nativeMapArgs(result, ::end)
            else -> super.onMethodCall(call, result)
        }
    }

    /**
     * Starts a live notification and returns its id.
     *
     * Takes [result] directly because [nativeMapArgs] reports every failure under the method name:
     * a type that isn't enabled has to arrive as [TYPE_NOT_REGISTERED_CODE] so the same mistake
     * reads the same way on iOS and Android. That check is not optional — the native
     * `startLiveNotification` mints an id whether or not the type is enabled, and
     * `LiveNotificationHandler` then drops the notification at debug level, so without it the caller
     * gets a real id and nothing ever renders.
     */
    @Suppress("UNCHECKED_CAST")
    private fun start(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<String, Any> ?: emptyMap()
        try {
            val module = requireNotNull(getPushModule()) { MODULE_UNAVAILABLE }
            val payload = requireNotNull(args.getAs<Map<String, Any>>("payload")) {
                "payload is required"
            }
            // Custom types have no built-in template, so they take the SDK's untyped map path and are
            // rendered by the app's own callback. Built-ins keep the typed path.
            val isCustom = payload.getAs<String>("type") == CUSTOM_TYPE_DISCRIMINATOR
            val activityType = if (isCustom) {
                requireCustomActivityType()
            } else {
                payload.requireString("type")
            }
            if (activityType !in enabledActivityTypes()) {
                return result.error(
                    TYPE_NOT_REGISTERED_CODE,
                    notRegisteredMessage(activityType),
                    null,
                )
            }
            val id = if (isCustom) {
                module.startLiveNotification(activityType, customData(payload))
            } else {
                module.startLiveNotification(parseData(payload))
            }
            result.success(id)
        } catch (ex: Throwable) {
            result.error(call.method, ex.localizedMessage, ex)
        }
    }

    private fun update(args: Map<String, Any>) {
        val module = requireNotNull(getPushModule()) { MODULE_UNAVAILABLE }
        val activityId = requireNotNull(args.getAs<String>("activityId")) {
            "activityId is required"
        }
        val payload = requireNotNull(args.getAs<Map<String, Any>>("payload")) {
            "payload is required"
        }
        if (payload.getAs<String>("type") == CUSTOM_TYPE_DISCRIMINATOR) {
            module.updateLiveNotification(activityId, requireCustomActivityType(), customData(payload))
        } else {
            module.updateLiveNotification(activityId, parseData(payload))
        }
    }

    /**
     * Ends a live notification. Dart may send a `payload` for iOS, where a final content-state is
     * what makes ActivityKit render a terminal state; Android renders its own terminal state (the
     * notification simply stops being ongoing), so the key is ignored here.
     */
    private fun end(args: Map<String, Any>) {
        val module = requireNotNull(getPushModule()) { MODULE_UNAVAILABLE }
        val activityId = requireNotNull(args.getAs<String>("activityId")) {
            "activityId is required"
        }
        module.endLiveNotification(activityId)
    }

    /**
     * Identifiers the SDK will actually render, read from the live push module config rather than
     * cached at init. The wrapper is not the only thing that can enable a type — an app (or a
     * generated config plugin) may call `enableLiveNotificationTypes` on its own builder — and a
     * cached copy would then be empty and refuse every start.
     */
    private fun enabledActivityTypes(): Set<String> =
        SDKComponent.pushModuleConfig.liveNotificationTypes

    /**
     * The app's own identifier for the custom template. Absent means the app sent a custom payload
     * without configuring `liveNotifications.customType` — say so, rather than starting a
     * notification the allowlist would silently drop.
     *
     * Falls back to the live config for the same reason [enabledActivityTypes] reads it: the custom
     * type may have been enabled without going through this wrapper. It is the one enabled
     * identifier that isn't a built-in template, and there is only ever one.
     */
    private fun requireCustomActivityType(): String = requireNotNull(
        customActivityType ?: enabledActivityTypes().firstOrNull { identifier ->
            LiveNotificationType.entries.none { it.identifier == identifier }
        },
    ) {
        "No custom Live Activity type is configured. Set `liveNotifications.customType` in your " +
            "Customer.io SDK config to your own reverse-DNS identifier, and render it from your " +
            "CustomerIOLiveNotificationsCallback."
    }

    /** Flattens the payload's `data` map. Android stringifies every value downstream anyway. */
    private fun customData(payload: Map<String, Any>): Map<String, Any> =
        payload.getAs<Map<String, Any>>("data") ?: emptyMap()

    private fun parseData(payload: Map<String, Any>): LiveNotificationData {
        return when (val type = payload.getAs<String>("type")) {
            LiveNotificationType.SEGMENTS.identifier -> LiveNotificationData.Segments(
                header = payload.requireString("header"),
                status = payload.requireString("status"),
                substatus = payload.getAs<String>("substatus"),
                segmentsTotal = payload.requireInt("segmentsTotal"),
                segmentsComplete = payload.requireInt("segmentsComplete"),
                trailingText = payload.getAs<String>("trailingText"),
            )

            LiveNotificationType.COUNTDOWN_TIMER.identifier -> LiveNotificationData.CountdownTimer(
                header = payload.requireString("header"),
                title = payload.requireString("title"),
                statusMessage = payload.getAs<String>("statusMessage"),
                endTime = payload.getAs<Number>("endTime")?.toLong(),
            )

            // A newer native SDK may know this type even though this wrapper build doesn't.
            // Reject softly (the caller turns this into a FlutterError) rather than crash.
            else -> throw IllegalArgumentException("Unsupported Live Activity template: $type")
        }
    }

    private fun Map<String, Any>.requireString(key: String): String =
        requireNotNull(getAs<String>(key)) { "$key is required" }

    private fun Map<String, Any>.requireInt(key: String): Int =
        requireNotNull(getAs<Number>(key)) { "$key is required" }.toInt()

    /**
     * No-op: Live Notification configuration is applied onto the FCM push module's config builder
     * via [applyLiveActivitiesConfig] rather than by adding a separate module.
     */
    override fun configureModule(builder: CustomerIOBuilder, config: Map<String, Any>) {}

    companion object {
        private const val PUSH_FCM_MODULE_NAME = "MessagingPushFCM"
        private const val MODULE_UNAVAILABLE =
            "Live Notifications are unavailable. Enable live activity templates in the SDK config."

        // Discriminator Dart sends for the custom template. Not a wire identifier — the notification
        // is started under the app's own `liveNotifications.customType`.
        private const val CUSTOM_TYPE_DISCRIMINATOR = "custom"

        // Mirrors the iOS handler's code so the same mistake reads the same way on both platforms.
        private const val TYPE_NOT_REGISTERED_CODE = "live_activity_type_not_registered"

        private fun notRegisteredMessage(activityType: String): String =
            "Live Activity type '$activityType' is not registered. Add it to " +
                "`liveNotifications.types` in your Customer.io SDK config (or set " +
                "`liveNotifications.customType` for a custom type), and make sure your app renders it."

        // The app's own identifier for the custom template, captured from config at SDK init. Only a
        // hint: `requireCustomActivityType` falls back to the live config, which is authoritative
        // however the type was enabled.
        @Volatile
        private var customActivityType: String? = null

        /**
         * App-provided callback used to render custom (app-defined) Live Notifications. The native
         * push callback can only be set at SDK build time, so the host app must register this
         * *before* initializing the Customer.io SDK from Dart; it is then applied onto the push
         * module's config in [applyLiveActivitiesConfig].
         */
        @Volatile
        private var liveNotificationCallback:
            io.customer.messagingpush.data.communication.CustomerIOLiveNotificationsCallback? = null

        /**
         * Registers the callback used to render custom Live Notifications. Must be called before the
         * Customer.io SDK is initialized from Dart.
         */
        @JvmStatic
        fun setLiveNotificationCallback(
            callback: io.customer.messagingpush.data.communication.CustomerIOLiveNotificationsCallback,
        ) {
            liveNotificationCallback = callback
        }

        /**
         * Applies live activity configuration onto the FCM push module's config builder. Live
         * Notifications are hosted by [ModuleMessagingPushFCM], so their config (enabled templates,
         * custom types, branding) is set on the same [MessagingPushModuleConfig].
         *
         * @param builder the push module's config builder.
         * @param config the `liveNotifications` config map from the customer app.
         */
        internal fun applyLiveActivitiesConfig(
            builder: MessagingPushModuleConfig.Builder,
            config: Map<String, Any>,
        ) {
            // Custom live notifications require the host app to render them; apply the callback the
            // app registered before SDK init (see [setLiveNotificationCallback]).
            liveNotificationCallback?.let { builder.setLiveNotificationCallback(it) }

            // Unrecognized identifiers are ignored: a newer native SDK may ship types this
            // wrapper build doesn't know, and that must never break the ones it does know.
            val templateTypes = config.getAs<List<*>>("types")
                ?.mapNotNull { it as? String }
                ?.mapNotNull { identifier ->
                    LiveNotificationType.entries.firstOrNull { it.identifier == identifier }
                }
                .orEmpty()
            if (templateTypes.isNotEmpty()) {
                builder.enableLiveNotificationTypes(*templateTypes.toTypedArray())
            }

            // The custom template. Allowlisting the identifier is both necessary and sufficient:
            // LiveNotificationHandler drops any push whose activityType isn't in this set, and a type
            // with no built-in template falls through to the host app's render callback.
            //
            // Assigned unconditionally: a re-initialize that drops `customType` must clear it, or
            // `start` would keep minting notifications under an identifier no longer allowlisted —
            // which the handler then discards, leaving the caller with an id and nothing on screen.
            val customType = config.getAs<String>("customType")?.trim()?.takeIf { it.isNotEmpty() }
            customActivityType = customType
            if (customType != null) {
                builder.enableCustomLiveNotificationTypes(customType)
            }

            val branding = config.getAs<Map<String, Any>>("branding")
            if (branding != null) {
                val accentColor = branding.getAs<String>("accentColorHex")
                    ?.let { runCatching { Color.parseColor(it) }.getOrNull() }
                    ?: Color.TRANSPARENT
                // A bundled drawable is preferred over a remote URL: it renders without a network
                // round-trip, so the logo is present on the very first frame.
                val logo = branding.getAs<String>("logoResource")
                    ?.let { name -> drawableResId(name)?.let(LiveNotificationAsset::Drawable) }
                    ?: branding.getAs<String>("logoUrl")?.let(LiveNotificationAsset::RemoteUrl)
                val smallIcon = branding.getAs<String>("smallIconResource")
                    ?.let { name -> drawableResId(name) }
                builder.setLiveNotificationBranding(
                    LiveNotificationBranding(
                        companyName = branding.getAs<String>("companyName").orEmpty(),
                        accentColor = accentColor,
                        smallIcon = smallIcon,
                        logo = logo,
                    ),
                )
            }
        }

        /**
         * Resolve a bundled drawable by name, or `null` when the host app doesn't ship one under
         * that name — a missing asset must degrade to "no image", never crash rendering.
         */
        private fun drawableResId(name: String): Int? {
            val context = SDKComponent.android().applicationContext
            return context.resources
                .getIdentifier(name, "drawable", context.packageName)
                .takeIf { it != 0 }
        }
    }
}
